src/webdemo-api.js

```diff
@@ -68,44 +71,68 @@ export default class WebDemoApi {
   /**
    * Request on execute Kotlin code.
    *
-   * @param code            - string
-   * @param compilerVersion - string kotlin compiler
-   * @param platform        - TargetPlatform
-   * @param args            - command line arguments
-   * @param theme           - theme of editor
-   * @param onTestPassed    - function will call after test's passed
-   * @param onTestFailed    - function will call after test's failed
+   * @param code                 - string
+   * @param compilerVersion      - string kotlin compiler
+   * @param platform             - TargetPlatform
+   * @param args                 - command line arguments
+   * @param theme                - editor theme
+   * @param onTestPassed         - a function that will be called if all tests pass
+   * @param onTestFailed         - a function that will be called if some tests fail (but after all tests are executed)
    * @param hiddenDependencies   - read only additional files
-   * @returns {*|PromiseLike<T>|Promise<T>}
+   * @param callback             - a callback for output chunks
    */
-  static executeKotlinCode(code, compilerVersion, platform, args, theme, hiddenDependencies, onTestPassed, onTestFailed) {
-    return executeCode(API_URLS.COMPILE, code, compilerVersion, platform, args, hiddenDependencies).then(function (data) {
-      let output = "";
-      let errorsAndWarnings = flatten(Object.values(data.errors));
-      let errors = errorsAndWarnings.filter(error => error.severity === "ERROR");
+  static executeKotlinCode(code, compilerVersion, platform, args, theme, hiddenDependencies, onTestPassed, onTestFailed, callback) {
+    const testResults = {
+      testsRun: 0,
+      totalTime: 0,
+      success: true
+    }
+    return executeCodeStreaming(API_URLS.COMPILE, code, compilerVersion, platform, args, hiddenDependencies, result => {
+      if (result.done) {
+        if (platform === TargetPlatform.JUNIT) {
+          const output = processJUnitTotalResults(testResults, onTestPassed, onTestFailed)
+          callback({
+            waitingForOutput: true,
```

Comment:
Is it good two callback updates for **JUNIT**?

Reply:
Two callback updates worked fine. But doing one update is probably better so I combined all fields in one call.